feat: take all neighbors into account when computing resource digests

otaviomacedo opened the pull request:
The current definition of the digest of a resource computes a hash of its properties plus the hashes of its dependencies. 

But if we have a graph like:

        A --> B
        C --> D

where `B` and `D` are identical, the current algorithm will compute the same digest for them, even if `A` is different from `C` (and therefore have different digests). This is undesirable, as some cases are considered ambiguous, even if we could very well tell them apart. Real world example:

<img width="326" height="182" alt="non-ambiguous" src="https://github.com/user-attachments/assets/32e442c6-9605-43b4-bdd5-a745f1a0426f" />

`Role1` and `Role2` tend to be identical, but they can nevertheless be told apart because at least the functions will have different properties. So, even if both roles are renamed, for example,  we can still provide a valid mapping.

This change improves the algorithm by also taking into account all neighbors of a resource for digest computation. It does this by doing a two pass over the graph: first navigating in topological order and computing the digests, then navigating in the opposite order and doing the same. So each resource has two digests, that are in turn, used to compute the final digest.


---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

otaviomacedo committed:
feat: take all neighbors into account when computing resource digests

rix0rrr commented:
So we now have two digests. Let's give them names (not the best ones, maybe you can come up with better ones).

* `downdigest` - the digest of a resource's properties and all of the resources it references.
* `updigest` - the digest of a resource's properties and all of the resources referencing it.

If we always use both digests to come up with the mapping digest (`mappingdigest = digest(down + up)`) we are sensitive to incoming references changing. Does that ever pose a problem?

Or would we be better off having a 2-step digest, where we map using the `downdigest` first, and then only involve the `updigest` if there is a collission?

otaviomacedo committed:
Using opposite graph only to resolve ambiguities

otaviomacedo commented:
> Typo? A is different from C ?

Yes, thanks. Updated.

> If we always use both digests to come up with the mapping digest (`mappingdigest = digest(down + up)`) we are sensitive to incoming references changing. Does that ever pose a problem?

Right now, this is not a problem because we require the graphs to be isomorphic anyway. But I went ahead and implemented your suggestion, so it's ready when we need it.
